checks/dangerous_workflow.go

```diff
@@ -0,0 +1,262 @@
+	}
+
+	return nil
+}
+
+func validateUntrustedCodeCheckoutRef(config map[interface{}]interface{}, path string,
+	dl checker.DetailLogger, pdata *patternCbData) error {
+	var jobs interface{}
+
+	// Now check if this is used with untrusted code checkout ref in jobs
+	jobs, ok := config["jobs"]
+	if !ok {
+		return nil
+	}
+
+	mjobs, ok := jobs.(map[string]interface{})
+	if !ok {
+		return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+	}
+
+	for _, value := range mjobs {
+		job, ok := value.(map[string]interface{})
+		if !ok {
+			return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+		}
+
+		if err := checkJobForUntrustedCodeCheckout(job, path, dl, pdata); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func checkPullRequestTrigger(config map[interface{}]interface{}) (bool, error) {
+	// Check event trigger (required) is pull_request_target
+	trigger, ok := config["on"]
+	if !ok {
+		return false, sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+	}
+
+	isPullRequestTrigger := false
+	switch val := trigger.(type) {
+	case string:
+		if strings.EqualFold(val, "pull_request_target") {
+			isPullRequestTrigger = true
+		}
+	case []string:
+		for _, onVal := range val {
+			if strings.EqualFold(onVal, "pull_request_target") {
+				isPullRequestTrigger = true
+			}
+		}
+	case map[interface{}]interface{}:
+		for k := range val {
+			key, ok := k.(string)
+			if !ok {
+				return false, sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+			}
+			if strings.EqualFold(key, "pull_request_target") {
+				isPullRequestTrigger = true
+			}
+		}
+	default:
+		return false, sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+	}
+	return isPullRequestTrigger, nil
+}
+
+func checkJobForUntrustedCodeCheckout(job map[string]interface{}, path string,
+	dl checker.DetailLogger, pdata *patternCbData) error {
+	steps, ok := job["steps"]
+	if !ok {
+		return nil
+	}
+	msteps, ok := steps.([]interface{})
+	if !ok {
+		return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+	}
+	// Check each step, which is a map, for checkouts with untrusted ref
+	for _, step := range msteps {
+		mstep, ok := step.(map[string]interface{})
+		if !ok {
+			return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+		}
+		// Check for a step that uses actions/checkout
+		uses, ok := mstep["uses"]
+		if !ok {
+			continue
+		}
+		muses, ok := uses.(string)
+		if !ok {
+			return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+		}
+		// Uses defaults if not defined.
+		with, ok := mstep["with"]
+		if !ok {
+			continue
+		}
+		mwith, ok := with.(map[string]interface{})
+		if !ok {
+			return sce.WithMessage(sce.ErrScorecardInternal, errInvalidGitHubWorkflow.Error())
+		}
+		// Check for reference. If not defined, uses default main branch.
```

Comment:
I don't think this is accurate for PR workflows. From https://github.com/actions/checkout#usage:
```
# The branch, tag or SHA to checkout. When checking out the repository that
# triggered a workflow, this defaults to the reference or SHA for that event.
# Otherwise, uses the default branch.
ref: ''
```
Sounds to be that when a PR triggers the workflow, then the latest commit in that PR will be checked out.

Reply:
How about tweaking the comment to "In the context of a `pull_request_target` event, the absence of `ref` checks out the target branch of the pull request" or something to this effect...?

Reply:
I'm trying to read up on this, but I believe actually it checks out the base of the PR which is trusted
(See this thread https://github.community/t/can-workflow-changes-be-used-with-pull-request-target/178626

and GITHUB_REF in https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#pull_request_target)

Reply:
Good catch. Im curious then: what if the base PR is cloned from a non-Branch-Protected branch on the original repo? Is it considered trusted?
Just curious... if there were testing workflows in that branch that auto-merge, etc it could be dangerous too.. probably not a very common threat anyway?

Reply:
Right -- I think that if the base branch wasn't branch protected that would be a problem as well. But I assume that you should be using BP check in conjunction

Reply:
I mean, main branch is protected, but branch used by the PR is not. Say:
- repo has branch `main` and `test`. `main` is protected, but `test` is not.
- attacker clone the repo and update `test` branch locally.
- attacker creates PR to merge its local cloned `test` into our `main` branch.

Is this situation possible? If so, is there a way to detect it?

Reply:
> attacker creates PR to merge its local cloned test into our main branch.

But then the target branch would be the main branch -- so the code being checked out in pull_request_target, GITHUB_REF, would be the main, not the untrusted code.

Reply:
ah, so I mis-understood the link. My original understanding was correct then. Sorry for the noise, ignore this comment :-)